onnxruntime/contrib_ops/cpu/quantization/moe_quantization_cpu.cc

```diff
@@ -272,111 +371,215 @@ Status QMoE::QuantizedMoEImpl(OpKernelContext* context,
-                const float* fc1_expert_bias_float = fc1_bias_float.get() + static_cast<int64_t>(SafeInt<int64_t>(expert_idx)) * fc1_bias_size;
-                for (int64_t i = 0; i < fc1_bias_size; ++i) {
-                  thread_fc1_output[i] += fc1_expert_bias_float[i];
-                }
+  if (is_decoding_scenario) {
+    // Decoding scenario: partition work by tokens * experts for better parallelization
+    // This allows multiple threads to work on different experts for the same token
+    // Cost estimate: 2 GEMM operations (FC1 + FC2) plus activation and bias operations
+    double cost_per_token_expert = static_cast<double>(moe_params.hidden_size * moe_params.inter_size * 2);  // Two GEMM operations per token-expert pair
+
+    concurrency::ThreadPool::TryParallelFor(
+        thread_pool, static_cast<std::ptrdiff_t>(work_units),
+        cost_per_token_expert,
+        [&](ptrdiff_t start_work, ptrdiff_t end_work) {
+          const int64_t thread_id = start_work / std::max<int64_t>(1, (work_units + num_threads - 1) / num_threads);
+
+          // Buffer management for decoding scenario
+          float* thread_base_buffer = thread_buffers.get() + thread_id * extended_buffer_size_per_thread;
+          float* thread_fc1_output = thread_base_buffer;
+          float* thread_fc2_output = thread_base_buffer + fc1_buffer_size_per_thread;
+          float* thread_accumulation_buffer = thread_base_buffer + total_buffer_size_per_thread;
+
+          // Process each token-expert work unit
+          for (std::ptrdiff_t work_idx = start_work; work_idx < end_work; ++work_idx) {
+            const int64_t token_idx = work_idx / moe_params.num_experts;
+            const int64_t expert_idx = work_idx % moe_params.num_experts;
+
+            // Check routing weight first to skip low-impact experts
+            const float routing_weight = router_probs_float_ptr[token_idx * moe_params.num_experts + expert_idx];
+            if (routing_weight <= 1e-4f) {
+              continue;  // Skip experts with negligible routing weight
+            }
+
+            const float* token_input = input_float_ptr + token_idx * moe_params.hidden_size;
+
+            // Initialize accumulation buffer for this expert's contribution
+            std::fill_n(thread_accumulation_buffer, static_cast<size_t>(moe_params.hidden_size), 0.0f);
+
+            // Process this expert for this token with optimized pointer access
+            ProcessSingleExpert(token_input, thread_fc1_output, thread_fc2_output, thread_accumulation_buffer,
+                                expert_idx, routing_weight, moe_params, is_swiglu,
+                                dequant_fc1_weights, dequant_fc2_weights,
+                                fc1_bias_float_ptr, fc2_bias_float_ptr, fc1_output_size);  // Atomically accumulate results to avoid race conditions
+            float* token_result = output_float_ptr + token_idx * moe_params.hidden_size;
```

Comment:
Data race bug.

For instance, Thread 1 might handle (token_idx=0, expert_idx=0) while Thread 2 handles (token_idx=0, expert_idx=1). Both threads will eventually attempt to write to the same memory location (output_float_ptr + token_idx * hidden_size) without any synchronization.